### PR TITLE
[feat] Layer Zero Relayer support

### LINKED ISSUE #17: Layer Zero Relayer support

### PR DESCRIPTION
**Pull request by ajimeno04**
This PR addresses the implementation of the LayerZero collector within the Relayer. Two scripts have been generated for this purpose:

One is the one that loads the configuration of and initialises a worker.
The code of the worker itself: it monitors two smart contracts to check that the lifecycle of transactions and crosschain messages has been completed successfully.

### CODE REVIEW

File: src/collector/layer-zero/layer-zero.worker.ts [hunk 1]
```diff
@@ -0,0 +1,617 @@
...
+    /**
+     * Main function to run the Layer Zero worker.
+     */
+    async run(): Promise<void> {
+        this.logger.info(
+            {
+                bridgeAddress: this.config.bridgeAddress,
+                receiverAddress: this.config.receiverAddress,
+            },
+            `Monitoring contracts: bridgeAddress and receiverAddress.`,
```
Comment (jsanmigimeno) [inline src/collector/layer-zero/layer-zero.worker.ts]:
```suggestion
            `LayerZero collector worker started.`,
```

File: src/collector/layer-zero/layer-zero.worker.ts [hunk 2]
```diff
@@ -0,0 +1,617 @@
...
+                        fromBlock = this.config.startingBlock;
+                    }
+                } else {
+                    fromBlock = this.currentStatus.blockNumber;
+                }
+            }
+            await wait(this.config.processingInterval);
+        }
+        const stopBlock = this.config.stoppingBlock ?? Infinity;
+        this.logger.info(`Stop block set to ${stopBlock}`);
```
Comment (jsanmigimeno) [inline src/collector/layer-zero/layer-zero.worker.ts]: Remove log.
```suggestion

```

File: src/collector/layer-zero/layer-zero.worker.ts [hunk 3]
```diff
@@ -0,0 +1,617 @@
...
+                if (
+                    this.config.maxBlocks != null &&
+                    blocksToProcess > this.config.maxBlocks
+                ) {
+                    toBlock = fromBlock + this.config.maxBlocks;
+                }
+                await this.queryAndProcessEvents(fromBlock, toBlock);
+                this.logger.info(
+                    { fromBlock, toBlock },
+                    `Scanning LayerZero Endpoint messages: fromBlock to toBlock.`,
```
Comment (jsanmigimeno) [inline src/collector/layer-zero/layer-zero.worker.ts]:
```suggestion
                    `Scanning LayerZero events.`,
```

File: src/collector/layer-zero/layer-zero.worker.ts [hunk 4]
```diff
@@ -0,0 +1,617 @@
...
+        toBlock: number,
+    ): Promise<void> {
+        const logs = await this.queryLogs(fromBlock, toBlock);
+        for (const log of logs) {
+            try {
+                await this.handleEvent(log);
+            } catch (error) {
+                this.logger.error(
+                    { log, error },
+                    `Failed to process event on layer-zero collector worker: log and error details.`,
```
Comment (jsanmigimeno) [inline src/collector/layer-zero/layer-zero.worker.ts]:
```suggestion
                    `Failed to process event on layer-zero collector worker.`,
```

File: src/collector/layer-zero/layer-zero.worker.ts [hunk 5]
```diff
@@ -0,0 +1,617 @@
...
+    }
+
+    /**
+     * Queries logs between two blocks.
+     * 
+     * @param fromBlock - The starting block number.
+     * @param toBlock - The ending block number.
+     * @returns A list of logs.
+     */
+    private async queryLogs(fromBlock: number, toBlock: number): Promise<Log[]> {
```
Comment (jsanmigimeno) [inline src/collector/layer-zero/layer-zero.worker.ts:255]: This logic is correct and works. However, please mirror the implementation of the `queryLogs` function of the `getter.worker.ts` that includes an endless retry mechanism until the query is succesful (i.e. the `while (logs == undefined)` loop).

File: src/collector/layer-zero/layer-zero.worker.ts [hunk 6]
```diff
@@ -0,0 +1,617 @@
...
+        let parsedLog: LogDescription | null = null;
+        if (log.address.toLowerCase() === this.bridgeAddress) {
+            parsedLog = this.layerZeroEnpointV2Interface.parseLog(log);
+        } else if (log.address.toLowerCase() === this.receiverAddress) {
+            parsedLog = this.receiveULN302Interface.parseLog(log);
+        }
+        if (parsedLog == null) {
+            this.logger.error(
+                { topics: log.topics, data: log.data },
+                `Failed to parse event: log topics and data details.`,
```
Comment (jsanmigimeno) [inline src/collector/layer-zero/layer-zero.worker.ts]:
```suggestion
                `Failed to parse LayerZero event.`,
```

File: src/collector/layer-zero/layer-zero.worker.ts [hunk 7]
```diff
@@ -0,0 +1,617 @@
...
+                break;
+
+            case 'PayloadVerified':
+                await this.handlePayloadVerifiedEvent(log, parsedLog);
+                break;
+
+            default:
+                this.logger.warn(
+                    { name: parsedLog.name, topic: parsedLog.topic },
+                    `Event with unknown name/topic received: parsedLog details.`,
```
Comment (jsanmigimeno) [inline src/collector/layer-zero/layer-zero.worker.ts]:
```suggestion
                    `Event with unknown name/topic received.`,
```

File: src/collector/layer-zero/layer-zero.worker.ts [hunk 8]
```diff
@@ -0,0 +1,617 @@
...
+                options,
+                sendLibrary
+            } = parsedLog.args as unknown as PacketSentEvent.OutputObject;
+            const packet = this.decodePacket(encodedPayload);
+            const srcEidMapped = this.layerZeroChainIdMap[packet.srcEid];
+            const dstEidMapped = this.layerZeroChainIdMap[packet.dstEid];
+
+            this.logger.debug(
+                { transactionHash: log.transactionHash, packet, options, sendLibrary },
+                'PacketSent event found: log details.',
```
Comment (jsanmigimeno) [inline src/collector/layer-zero/layer-zero.worker.ts]:
```suggestion
                'PacketSent event found.',
```

File: src/collector/layer-zero/layer-zero.worker.ts [hunk 9]
```diff
@@ -0,0 +1,617 @@
...
+                            blockHash: log.blockHash,
+                            transactionHash: log.transactionHash,
+                        },
+                        log.transactionHash,
+                    );
+
+                    this.logger.info(
+                        { transactionHash: log.transactionHash },
+                        'Primary AMB message created using setAmb: transactionHash details.',
+                    );
```
Comment (jsanmigimeno) [inline src/collector/layer-zero/layer-zero.worker.ts]: Make use of the following log statement to be consistent with the Wormhole collector implementation. Also, include the `payloadHash` on the log, and move the full log statment after the `await this.store.setPayload(...)` call. This way we only emit 1 log per collected message.
```suggestion
                    this.logger.info(
                        {
                            messageIdentifier: decodedMessage.messageIdentifier,
                            transactionHash: log.transactionHash,
                            payloadHash
                        },
                        'Collected message.',
                    );
```

File: src/collector/layer-zero/layer-zero.worker.ts [hunk 10]
```diff
@@ -0,0 +1,617 @@
...
+                    await this.store.setPayload('layer-zero', 'ambMessage', payloadHash, {
+                        messageIdentifier: decodedMessage.messageIdentifier,
+                        destinationChain: dstEidMapped,
+                        payload: encodedPayload,
+                    });
+
+                    this.logger.info(
+                        { payloadHash, transactionHash: log.transactionHash },
+                        'Secondary AMB message created with payload hash as key using setPayloadLayerZeroAmb: payloadHash and transactionHash details.',
+                    );
```
Comment (jsanmigimeno) [inline src/collector/layer-zero/layer-zero.worker.ts]: Remove this log statement.
```suggestion
```

File: src/collector/layer-zero/layer-zero.worker.ts [hunk 11]
```diff
@@ -0,0 +1,617 @@
...
+                } catch (innerError) {
+                    this.logger.error(
+                        { innerError, log },
+                        'Failed to process specific sender packet: innerError and log details.',
+                    );
+                    throw innerError;
+                }
```
Comment (jsanmigimeno) [inline src/collector/layer-zero/layer-zero.worker.ts]: Remove this catch statement, only keep the 'outer' one.

File: src/collector/layer-zero/layer-zero.worker.ts [hunk 12]
```diff
@@ -0,0 +1,617 @@
...
+            } else {
+                this.logger.debug(
+                    { sender: packet.sender },
+                    'Skipping packet: sender is not a GARP contract.',
+                );
+            }
+        } catch (error) {
+            this.logger.error({ error, log }, 'Failed to handle PacketSent event: error and log details.');
```
Comment (jsanmigimeno) [inline src/collector/layer-zero/layer-zero.worker.ts]:
```suggestion
            this.logger.error({ error, log }, 'Failed to handle PacketSent event.');
```

File: src/collector/layer-zero/layer-zero.worker.ts [hunk 13]
```diff
@@ -0,0 +1,617 @@
...
+    private async handlePayloadVerifiedEvent(
+        _log: Log,
+        parsedLog: LogDescription,
+    ): Promise<void> {
+        const { dvn, header, confirmations, proofHash } = parsedLog.args as any;
+        const decodedHeader = this.decodeHeader(header);
+        const srcEidMapped = this.layerZeroChainIdMap[decodedHeader.srcEid];
+        const dstEidMapped = this.layerZeroChainIdMap[decodedHeader.dstEid];
+        if (srcEidMapped === undefined || dstEidMapped === undefined) {
+            throw new Error('Failed to map srcEidMapped or dstEidMapped.');
```
Comment (jsanmigimeno) [inline src/collector/layer-zero/layer-zero.worker.ts]: Make this a log statement, and simply return from the function without erroring. i.e. `this.logger.error({ srcEid, dstEid }, 'Failed to map ...`

File: src/collector/layer-zero/layer-zero.worker.ts [hunk 14]
```diff
@@ -0,0 +1,617 @@
...
+        }
+        if (
+            decodedHeader.sender.toLowerCase() ===
+            this.incentivesAddresses[srcEidMapped]
+        ) {
+            this.logger.info(
+                { dvn, decodedHeader, confirmations, proofHash },
+                'PayloadVerified event decoded: dvn, decodedHeader, confirmations, and proofHash details.',
```
Comment (jsanmigimeno) [inline src/collector/layer-zero/layer-zero.worker.ts]:
```suggestion
                'PayloadVerified event decoded.',
```

File: src/collector/layer-zero/layer-zero.worker.ts [hunk 15]
```diff
@@ -0,0 +1,617 @@
...
+                this.logger.error(
+                    { proofHash },
+                    'No data found in database for the given payloadHash: proofHash details.',
+                );
+                return;
+            }
+            this.logger.info(
+                { payloadData },
+                'Data fetched from database using payloadHash: payloadData details.',
+            );
```
Comment (jsanmigimeno) [inline src/collector/layer-zero/layer-zero.worker.ts]: Remove this log statement. It is implied that the data was loaded correctly if there is no error printed to the logs.
```suggestion
```

File: src/collector/layer-zero/layer-zero.worker.ts [hunk 16]
```diff
@@ -0,0 +1,617 @@
...
+                    dvn,
+                    decodedHeader.dstEid,
+                );
+                const isVerifiable = await checkIfVerifiable(
+                    this.receiveULN302,
+                    config,
+                    keccak256(header),
+                    proofHash,
+                );
+                this.logger.info({ dvn, isVerifiable }, 'Verification result checked: dvn and isVerifiable details.');
```
Comment (jsanmigimeno) [inline src/collector/layer-zero/layer-zero.worker.ts]: Remove this log statement. Make a new one only for the case in which `!isVerifiable` (i.e. put it as an `else` case on the following `if`) , and log that the payload is not verifiable using the `debug` level
```suggestion
```

File: src/collector/layer-zero/layer-zero.worker.ts [hunk 17]
```diff
@@ -0,0 +1,617 @@
...
+                if (isVerifiable) {
+                    const ambPayload: AmbPayload = {
+                        messageIdentifier: '0x' + payloadData.messageIdentifier,
+                        amb: 'layer-zero',
+                        destinationChainId: dstEidMapped.toString(),
+                        message: payloadData.payload,
+                        messageCtx: '0x',
+                    };
+                    this.logger.info({ proofHash }, `LayerZero proof found: proofHash details.`);
```
Comment (jsanmigimeno) [inline src/collector/layer-zero/layer-zero.worker.ts]:
```suggestion
                    this.logger.info({ proofHash }, `LayerZero proof found.`);
```

File: src/collector/layer-zero/layer-zero.worker.ts [hunk 18]
```diff
@@ -0,0 +1,617 @@
...
+                    await this.store.submitProof(
+                        this.layerZeroChainIdMap[decodedHeader.dstEid]!,
+                        ambPayload,
+                    );
+                }
+            } catch (error) {
+                this.logger.error(
+                    { error: tryErrorToString(error) },
+                    'Error during configuration verification: error details.',
```
Comment (jsanmigimeno) [inline src/collector/layer-zero/layer-zero.worker.ts]:
```suggestion
                    'Error during payload verification.',
```

File: src/collector/layer-zero/layer-zero.worker.ts [hunk 19]
```diff
@@ -0,0 +1,617 @@
...
+        };
+        const isVerifiable = await receiveULN302.verifiable(
+            formatConfig,
+            headerHash,
+            payloadHash,
+        );
+        return isVerifiable;
+    } catch (error) {
+        console.error('Error verifying the configuration: ', error);
+        throw new Error('Error verifying the configuration: error details.');
```
Comment (jsanmigimeno) [inline src/collector/layer-zero/layer-zero.worker.ts]: Remove the `console.error` and update the error statement. This is not ideal, as the `verifiable()` call is only given one chance to succeed, but let's move forward for now.
```suggestion
        throw new Error('Error verifying the payload.');
```

**Review by jsanmigimeno — CHANGES_REQUESTED** [review]
Looks good, only minor changes requested (mostly to the log statements).

File: src/collector/layer-zero/layer-zero.worker.ts (hunk 11 above)
Comment (ajimeno04) [reply src/collector/layer-zero/layer-zero.worker.ts]: I'm not sure waht do you mean by outer one. Take a look to check if I got it right.

File: src/collector/layer-zero/layer-zero.worker.ts (hunk 19 above)
Comment (ajimeno04) [reply src/collector/layer-zero/layer-zero.worker.ts]: Check on this if I solved as expected

File: src/collector/layer-zero/layer-zero.worker.ts (hunk 16 above)
Comment (ajimeno04) [reply src/collector/layer-zero/layer-zero.worker.ts]: Check of how I resolved this and if I got the right idea

File: src/collector/layer-zero/layer-zero.worker.ts (hunk 11 above)
Comment (jsanmigimeno) [reply src/collector/layer-zero/layer-zero.worker.ts]: Yes, it is right.

**Review by jsanmigimeno — APPROVED** [review]

**Pull request merged**